### ISSUE TITLE
Special treatment of pre-installed packages by the solver

### ISSUE THREAD

**Issue by hasufell** [haskell/cabal#9669]
The cabal solver seems to treat pre-installed packages specially (e.g. those shipped with GHC).

To reproduce:

```
git clone https://github.com/hasufell/toto.git
cd toto
ghcup run --ghc 9.4.8 cabal build
```

This should cause a failure, because ghc-9.4.8 ships with filepath-1.4.2.2, but the package above uses modules from 1.4.100.1. The package has no upper bounds on filepath. **For any other package non-pre-installed package, the solver would pick the latest**.

I understand that this is by design, but I question this design here, because:

* it makes it harder for core library maintainers to ship bugfixes
* it's a potential security risk

@mpickering found out that there used to be a `--upgrade-dependencies` switch, which is now disabled.

I argue that the **default should be to pick the latest possible version anyway**.

----

CCing some potentially interested parties: @simonpj @frasertweedale

**Comment by mpickering** [comment]
Branch which re-enables `--upgrade-dependencies` (in a unfinished way) - https://github.com/mpickering/cabal/tree/wip/upgrade-dependencies

It seems sensible to me to choose pre-installed packages, then you don't have to build them again. If the version constraints don't disallow it then the solver could choose that install plan anyway. If you really don't want a package to be part of the install plan then perhaps instead you want a means to instruct the solver to never choose a particular version as an additional constraint form.

**Comment by Mikolaj** [comment]
Regardless of the outcome of the dicussion, it would make sense to find the PR that commented `--upgrade-dependencies` out, understand how this happened and prevent for the future, e.g., by guarding this functionality with tests and also by documenting it (better).

**Comment by hasufell** [comment]
@mpickering I'm worried about defaults here. I don't think there's an easy way to tell my users that there was a subtle bug in `filepath` in `splitFileName`. No one reads the changelog. GHC already ships the version.

What if this is a security bug? What do I do? I have no communication channels. A `cabal update && cabal build` should leave your project in the best possible state (bugfixes, security fixes) without further interventions/constraints required by the end users.

This is how all linux distributions work, afaik. "Saving compilation time" seems like a questionable priority, imo (as a default).

**Comment by Mikolaj** [comment]
@mpickering also says about `--upgrade-dependencies`: "There is the `-prefer-oldest` option already and this I suppose is a `--prefer-newest` ... decide how it should interact with `--prefer-oldest`".

**Comment by Mikolaj** [comment]
CC @grayjay, @gbaz

**Comment by ulysses4ever** [comment]
I agree that --prefer-newest (so, no special-casing the boot packages) would be a cleaner default. Perhaps, there should be --prefer-installed to get the current default if we change the default to --prefer-newest?

**Comment by Mikolaj** [comment]
While I like the default @hasufell proposed better, for the reasons give and also because it's more uniform, I worry it would hit hard the users with big sets of installed packages, which may include Nix users, Linux distribution users and v1-/cabal-env users (whether for teaching purposes or others). So we'd need some good backward compatibility scheme.

**Comment by Mikolaj** [comment]
Another solution, which unfortunately increases complication, might be to treat libraries installed together with GHC (and perhaps all installed not by the user directly, but by install/upgrade scripts) specially and apply `--upgrade-dependencies` to them, while keeping user-installed libraries immutable, on the premise that the user knows what the user is doing (and that installing libraries is rare and discouraged).

**Comment by hasufell** [comment]
CCing @Ericson2314 @angerman wrt Nix

**Comment by michaelpj** [comment]
> I don't think there's an easy way to tell my users that there was a subtle bug in filepath in splitFileName. No one reads the changelog. GHC already ships the version.

What happens if you deprecate the version with the bug (the installed version)? Will cabal still prefer it?

**Comment by gbaz** [comment]
I'm a bit confused. In fact, the way we use nix at work involves "pre-installing" everything in the sense that everything goes into a package database which nix provides, instead of the cabal store, no?

So with the current behavior, if I am building `foo` which depends on `bar-1` and the latter is in the package database nix provides, then even if `bar-1.1` is released, `cabal build` will still use `bar-1`. With the new behavior, `cabal-build` would download and build `bar-1.1` even though the nix configuration specified `bar-1`.

(the workaround for this, which is possible but slightly irritating, is to use a cabal.project or the like to _disable_ hackage or any other package repository for packages developed in a nix provided environment)

**Comment by gbaz** [comment]
All that said, I modestly prefer the current behavior, in part because I'm afraid of changing this sort of stuff given large and unpredictable effects it may have on many users, and in part because users will expect that if they have a `containers` that works installed, then cabal will just use it.

I do think an explicit flag like `upgrade-dependencies` (though that is a terrible name, given its semantics, and `prefer-newest` or the like is better) is useful, to make this behavior more controllable.

**Comment by hasufell** [comment]
I don't think cabal has any obligation to "not break nix". It's nix packagers obligation to keep it working. 

Changes like the proposed one would be communicated early enough with a migration period, so that users can adapt and opt out of the changed behavior. 

This is the same with the v1 vs v2 change that the cabal team executed over several years. Except this one seems **much less** disruptive.

I can't see how the current behavior is a sensible default from any angle, if it causes average users to miss bugfixes. It is not safe.

**Comment by Ericson2314** [comment]
@gbaz What I mean is that in Haskell.nix-style approaches *planning* takes place with no / empty package database. Ideally even for `cabal build` / in Nix shell usage, we'd still use the original pure / *ex nihilo* plan.

The current trick of "re-planning" in the Nix shell and hoping it solves for as little few possible not-already-built things as possibly is comparatively gross, and (IIRC) runs into issues when sources are funny (e.g. modified local packages).

(That said, the "already installed" constraint is useful for the above hack, and I imagine also useful for anyone that is wondering why their boot packages aren't being used under this issue's proposal.)

**Comment by frasertweedale** [comment]
> @mpickering I'm worried about defaults here. I don't think there's an easy way to tell my users that there was a subtle bug in `filepath` in `splitFileName`. No one reads the changelog. GHC already ships the version.
> 
> What if this is a security bug? What do I do? I have no communication channels. A `cabal update && cabal build` should leave your project in the best possible state (bugfixes, security fixes) without further interventions/constraints required by the end users.
> 
Speaking as a member of the Haskell Security Response Team, our hope is that cabal-install will be enhanced to directly use the data from the advisory database, and either omit affected packages from build plans by default, or alert users when build plans contain affected packages.

This issue poses some good questions but I don't think the SRT would have an opinion on it one way or the other, given the objective of more explicit cabal-install features/behaviour regarding known security issues.

**Comment by Bodigrim** [comment]
The issue with forcing as much newest dependencies as possible is that your library/app might use a very different set of dependencies than your tests. If tests involve [`doctest`](https://hackage.haskell.org/package/doctest), which is quite common, their build plan includes `ghc`-the-package and so sets in stone all boot libraries as shipped with GHC. The lib/app most likely does not depend on `ghc` and would be free to build against latest and greatest. Overall effect would be that you are testing not what you are shipping. 

I think making `ghc` reinstallable would be an important stepping stone.

**Comment by hasufell** [comment]
@simonpj 

> I am baffled about why we could possibly want to treat wombat and filepath differently, just because filepath happens, through some accident of fate, to come with GHC.

This is mostly correct, with the caveat that cabal treats *any* package that is in the **global package db** specially. It just so happens that `cabal v2-build` and `cabal v2-install` (which are now default) don't touch the global package db anymore as opposed to `cabal v1-install` (legacy). So for most users, the global package db just contains what GHC ships with.

That, imo, makes it even worse. There are many other mechanisms to avoid cabal rebuilds (e.g. just don't run `cabal update`, use a freeze file, pass certain flags to cabal, ...). For *saving space* it seems totally backwards and what we actually want there is: https://github.com/haskell/cabal/issues/3333

**Comment by hasufell** [comment]
@Bodigrim 

> If tests involve [doctest](https://hackage.haskell.org/package/doctest), which is quite common, their build plan includes ghc-the-package and so sets in stone all boot libraries as shipped with GHC.

I understand doctest is a special case, but I do not believe this justifies having the current default.

> Overall effect would be that you are testing not what you are shipping.

It is the maintainers responsibility to ensure testing across multiple setups. If they can't do that, then their cabal version bounds are *simply wrong* or their test suite is sub-par.

For anyone wondering, there are two practical solutions to avoid doctest:

* don't use doctest, but generate tests yourself from comments: https://github.com/haskell/filepath/blob/master/Generate.hs
* use [cabal-docspec](https://github.com/phadej/cabal-extras/blob/master/cabal-docspec/MANUAL.md), which does not depend on `ghc` library, but ghci

**Comment by hasufell** [comment]
@frasertweedale 

> Speaking as a member of the Haskell Security Response Team, our hope is that cabal-install will be enhanced to directly use the data from the advisory database, and either omit affected packages from build plans by default, or alert users when build plans contain affected packages.

I don't want to digress too much on this, but I'm rather surprised by this sentiment and disagree rather strongly (wrt this being enough).

I'll put my response in a collapsible section to keep the thread clean. I'm happy to continue that discussion privately or on the security response team issue tracker.

<details>
  <summary>Wrt security updates</summary>

To my knowledge, there are currently two main definitions about "software security" (or "insecurity").

1. "unexpected computation", which was first prominently described in several [langsec papers](https://langsec.org/) (e.g. "Security Applications of Formal Language Theory" and "The Halting Problems of Network Stack
Insecurity") and for which there exist very few practical models.
2. model security risk based on market means (as in: how expensive is it for an attacker to crack a system), e.g. "Computer Security Strength & Risk:A Quantitative Approach" by Stuart Edward Schechter

For the first definition there are a variety of technical models, but most of them are rather hard to apply. E.g. there is "An attack Surface Metric" by Pratyusa K. Manadhata, which attempts to model a system based on IO automatons and syscalls a user can trigger.
Several LANGSEC papers suggest to treat program input as a *protocol* and start with a strict parser and then model the system accordingly to understand possibly system states.
Several MAC/data-flow approaches have been developed to aid with the relationship of user input and control flow (also in Haskell). But these are very specific techniques that a *software engineer* applies.

The second definition allows more insights into the whole business of redistribution, software maintenance, supply chain issues, buying 0-day exploits or DDoS attacks on the darkweb, etc.
We assume that it is possible to crack any system, depending on available energy and assets. As such, in order to protect from attacks, **we want to make attacking the system more expensive**.
This can be achieved by a number of things: strong cryptographic protocols, network architecture, etc.

But most importantly: packaging and update policies. The easiest way to make attacks more expensive is **to always update to the latest versions of all packages**. The reason is that an attacker has to invest more time in studying/keeping up with new versions of software package and coming up with attacks (or buying them somewhere) as compared to versions that have been around for 2 years. There has just been less time to adapt. Software updates are disruptive for attackers. In that way, the most secure model is a "rolling release distro".

In addition, there is no clear definition in research of what a "security bug" is and this idea has also been rejected numerous times by the Linux kernel (see various interviews and ML posts by Greg Kroah-Hartman, the maintainer of the stable linux branch). The Linux kernel backports any bugfix. Because any bug can (as per the "unexpected computation" definition of insecurity) potentially lead to a security vulnerability, even if none is known yet. You are **not** safe just because you're not running a version that has no publicly known vulnerabilities.

All this said... we want to optimize an ecosystem of packages in a way that an attacker has to invest *a lot of resources* to drive attacks on users. And the most important way to do this is to be as aggressive as possible with software updates, regardless of CVEs. This conflicts (heavily) with the current default of cabal.

</details>

**Comment by Mikolaj** [comment]
@simonpj:

> I install a package wimwam using cabal, and it turns out that cabal's build plan installs the dependency wombat-2.7.2

I think modern cabal works rather differently. The new cabal v2-build (or v2-install) builds wombat-2.7.2 locally and does not "install" it, at least not in the same sense that GHC distribution installs the packages it comes with. Modern cabal discourages installing any libraries and encourages building them anew (with smart caching via "store").

In fact, if GHC stopped providing/exposing the bundled packages, the problem of the exceptional treatment of installed packages would be immediately gone (until the user insists on manually installing some other packages, which is discouraged and hard to do properly). If GHC ships with the packages so that the user saves on compilation, it's no wonder cabal tries to accommodate it. However, I'm guessing GHC exposes the packages, because the `ghc` package (or any non-reinstallable package?) depends on them and `ghc` can't be re-built/reinstalled/relinked (in particular, to depend on different versions/builds of its dependencies). See https://github.com/haskell/cabal/issues/9064#issuecomment-1609929980 and many related issues. I'm sure @bgamari or @mpickering could easily confirm or deny.

Therefore, the inconsistent cabal behaviour may be caused primarily by `ghc` and others not being reinstallable/rebuildable (and secondarily, by attempting backward compatibility for old v1-/Setup workflows, such as Nix, Linux distros, old setups for Haskell courses where each student is supposed to have the same versions of dependencies and freeze files were not yet a thing). If so, we can wait until `ghc`/others are reinstallable and then the problem vanishes (unless the user introduces it independently). Or we could try to limit the special cabal behaviour to build plans that include `ghc`, but then such build-plans are treated specially. If `template-haskell` is another case of a non-reinstallable package that depends on reinstallable packages (I remember rebuilding it in the past, but that's no longer possible, I think?), this makes the specially treated build plans much more common and harder to describe succinctly to the user.

Am I anywhere close to the root cause of keeping this old functionality in modern cabal? Can cabal handle GHC in some alternative way without incurring this irregular behaviour? E.g., what can go wrong if GHC renames all the packages (in the package db and/or on Hackage) it bundles so that they can't be reinstalled at all?

**Comment by simonpj** [comment]
> Therefore, the inconsistent cabal behaviour may be caused primarily by ghc and others not being reinstallable/rebuildable 

Yes indeed:
* IF the build plan uses `ghc-the-package` 
* THEN you are stuck with particular versions of the packages that `ghc-the-package` depends on

But that is a simple consequence of depending on `ghc-the-package`, which in turn depends on a particular wired-in version of `filepath`.

But let's **suppose that  your build plan does *not* depend on `ghc-the-package` or `template-haskell`** (a very common case).  Now `filepath` has no constraints -- cabal is entirely free to rebuild it locally.  The fact that there is a pre-installed version is irrelevant, no?   So my question remains: why is `filepath` (and other packages that happen to come with GHC) treated specially?

**Comment by michaelpj** [comment]
> Therefore, the inconsistent cabal behaviour may be caused primarily by ghc and others not being reinstallable/rebuildable 

I think this is not quite right. I don't think anyone is asking to change the behaviour when you depend on a non-reinstallable package: there we really do have to go with what GHC ships, and that's that.

I think the request is about packages that _are_ reinstallable, but happen to have versions in the global package-db, like `filepath`. Then the preference seems less justifiable.

**Comment by mpickering** [comment]
The reason that things in the `GlobalPackageDB` are treated specially is because of the definition of `corePackageDbs` in `Distribution.Client.ProjectPlanning`. This only looks in `GlobalPackageDb` and any extra package databases that a user has configured. 

In general, it is a bit of an issue that `cabal-install` and `Cabal` assume anything about the structure of package databases (see #3728). There are many assumptions baked into both projects that you want to use the global package database and the things in there are privileged. This is primarily a legacy from the old days when `GHC` was much less flexible about being able to specify a package database stack, but now it is completely agnostic.

**Comment by Mikolaj** [comment]
> why is filepath (and other packages that happen to come with GHC) treated specially?

> Then the preference seems less justifiable.

My guess is that cabal covers the case of dependencies of non-reinstallable package in a lazy way --- by treating specially all packages that reside in the relevant package DB. This is has several advantages: simplicity of implementation, simplicity of configuration (though, as @mpickering states, this may be too hard-wired at this point), an extra benefit of backward compatibility for other legacy workflows using a central package DB, simplicity of conveying the behaviour to the user (though it's probably not conveyed yet or not well enough).

Which is why I'm considering the other option, changing the behaviour of GHC, not of cabal. Or even of the GHC installer, e.g., making it rename all the packages it installs. That's probably absurd for fundamental reasons, but I'd like to improve my understanding of the situation by learning why exactly.

**Comment by gbaz** [comment]
The comments here are swaying me towards supporting a change here. However, I really do feel it needs to be flag-controlled and I'm definitely worried that any deep change like this could well confuse workers and disrupt workflows in a way that is very unexpected and hard to diagnose, especially for those who don't read release notes.

**Comment by Ericson2314** [comment]
That sounds fine to me. It's definitely a behavior change, quite arguably a breaking one. A new flag, switch default, etc. deprecation cycle should be fine.

**Comment by Mikolaj** [comment]
OK, let's assume we want the change and the minimal goal is to let users ensure they get the latest Hackage versions of all reinstallable packages without the need to edit any bounds anywhere. 

Let's brainstorm and list all options we have, from the least likely to cause breakage to the most likely. I will update the list as proposals and facts emerge (I'm quite clueless).

1. Bring back `--upgrade-dependencies`, document it, guard it against removal with tests.
2. Or install the packages GHC comes with to a different package db, one for which cabal does not hesitate to upgrade any packages.
3. Or don't upgrade a package by default only if the install plan contains a non-reinstallable package that depends on it (plus the backward compat measures from the next point).
4. Or make the `--upgrade-dependencies` behaviour the default, provide the opposite flag, warn one major release early.

Any more ideas?

Is 3 and 4 the same? Is 2 possible at all? Once we stop brainstorming, but not before, let's focus on drawbacks and benefits.

**Comment by hasufell** [comment]
There's one more subtlety that can be confusing (and is in fact an argument for removing the special treatment).

GHC sometimes ships libraries with different *cabal flag* configurations than you would get installing straight from hackage. As @bgamari notes [here](https://github.com/haskell/text/issues/487#issuecomment-1382245441) `text-2.0` shipped with GHC has `simdutf` flag disabled, but the [version on hackage](https://hackage.haskell.org/package/text-2.0) has it enabled by default.

This can be confusing as witnessed here: https://github.com/haskell/text/issues/487

**Comment by michaelpj** [comment]
Personally I would prefer a positive flag for preferring flags in the global package db, maybe `--prefer-globally-installed`, which could be on by default with a deprecation cycle to change the default. `--upgrade-dependencies` does not jump out to me as related to this issue!

> GHC sometimes ships libraries with different cabal flag configurations than you would get installing straight from hackage

I think this is just https://github.com/haskell/cabal/issues/8702 ? Although if that was fixed I'm unsure what we would expect to happen? Maybe:
- The solver makes a flag selection for `text`, which in the absence of user input would be `-simdutf`
- Cabal notices that the globally installed `text` does not match that flag selection so doesn't use it
- Therefore Cabal will essentially never use the globally installed `text` (which might be fine!)

**Comment by phadej** [comment]
> Cabal notices that the globally installed text does not match that flag selection so doesn't use it

Cabal cannot notice the flag selection of installed packages: the flags are not recorded in IPI.

**Comment by mpickering** [comment]
There seems to be some suggestion in https://discourse.haskell.org/t/is-cabal-install-stable-enough-yet-that-ghc-should-be-inflicting-it-on-newbies/9979/52 that `cabal install --lib` will add pre-installed packages (and hence be relevant to this ticket).

However, to my understanding that is false because `v2-` commands don't consult the user package database (see definition of `corePackageDbs` in Distribution.Client.ProjectPlanning. Therefore this default only affects the global package db and any package database specified to cabal with the `--package-db` flag. 

If I am mistaken about this then a reproduce which demonstrates the issue would be much appreciated.

**Comment by hasufell** [comment]
> If I am mistaken about this then a reproduce which demonstrates the issue would be much appreciated.

There's a clear reproducer for the original topic:

> To reproduce:
> 
> ```
> git clone https://github.com/hasufell/toto.git
> cd toto
> ghcup run --ghc 9.4.8 -- cabal build
> ```
> 
> This should cause a failure, because ghc-9.4.8 ships with filepath-1.4.2.2, but the package above uses modules from 1.4.100.1. The package has no upper bounds on filepath. **For any other non-pre-installed package, the solver would pick the latest**.
> 
>

----

> that cabal install --lib will add pre-installed packages (and hence be relevant to this ticket).

I'm not sure if anyone suggested that and it seems out of scope of this ticket how `install --lib` behaves. It is broken anyhow.

This ticket is about the solver.

**Comment by mpickering** [comment]
It seems that I misunderstood that suggestion in the thread (it is quite complicated all the discussions going on). Thank you for clarifying that. 

In particular the comment "Do v2 builds even look at the globally installed packages?", globally installed packages might mean, packages installed in the the global package database OR packages installed in the user package database by `cabal install --lib`.

**Comment by TeofilC** [comment]
What's the status of this?


> Personally I would prefer a positive flag for preferring flags in the global package db, maybe `--prefer-globally-installed`, which could be on by default with a deprecation cycle to change the default. `--upgrade-dependencies` does not jump out to me as related to this issue!

I think @michaelpj's suggestion is great. Though I think it's a bit tempting to just make the entire change in one release. I struggle to see how we could communicate this "deprecation" to users, and I imagine we won't get much usage until it becomes the default -- I think the users who would test this feature would just as easily test it from a nightly build.

---

I would be really keen to make progress here because I think this is one of the issues blocking decoupling GHC upgrades from boot library upgrades, and I think that would be a huge improvement.

**Comment by phadej** [comment]
> though maybe I'm misunderstanding this issue?)

This issue is about *default* behavior. `cabal-install` would solve for newer version and pick one if forced to. So no, this issue is not blocking anything AFAICT.

**Comment by TeofilC** [comment]
> This issue is about _default_ behavior. `cabal-install` would solve for newer version and pick one if forced to. So no, this issue is not blocking anything AFAICT.

Good point. I should've been more precise. Here is the sort of scenario I had in mind:

(Assuming the package uses cabal-install and doesn't depend on `ghc:lib`)

If I want to submit a PR to bump the bound for a non-boot library, I can just update the bound, and be confident that it will  be tested by CI.

If I want to bump a boot library bound, then I need to apply this workaround where I forbid the old version. I think most of the community is understandably not aware that this is even possible. Even if someone was aware, I think most maintainers wouldn't want to deal with this extra complexity.  They are likely to instead wait for a version of GHC where that library is bundled.

**Comment by TeofilC** [comment]
@Mikolaj wrote early on in the discussion:
> Regardless of the outcome of the discussion, it would make sense to find the PR that commented `--upgrade-dependencies` out, understand how this happened and prevent for the future, e.g., by guarding this functionality with tests and also by documenting it (better).

Git blame points us to this commit by @dcoutts https://github.com/haskell/cabal/commit/324b3240ff3bd4eed2d33a01203edbf28dcea214#diff-e2de3403daa75f77ddd177d0a040f0547097abddb360328293ba46da21673a4e

And as far as I can tell `--upgrade-dependencies` is introduced here in the PR that implemented nix-style builds already commented out with this comment:
```
       -- Things that only make sense for manual mode, not --local mode
       -- too much control!
```

So it sounds like these have never been applicable to v2 style commands

**Comment by phadej** [comment]
`--upgrade-dependencies ` is a legacy from `v1-build`.

The `v2` commands have a lot of flags which are inherited from `v1-build` and do nothing or don't make sense, because the `v2` interface wasn't started from scratch, but partially reuses `v1-build` command definitions.

You correctly *blame* Duncan. It might made sense to get something quickly done, but in long term that was a bad choice. The `v2` cli interface is not clean.

`v1-build` used a single package database, so `--upgrade-dependencies` was a thing.

**Comment by phadej** [comment]
> be confident that it will be tested by CI.
> If I want to bump a boot library bound, then I need to apply this workaround where I forbid the old version.

Well... you have to keep testing against the old versions too, because in larger setup there might be `ghc`. `--prefer-oldest` is a recent addition, which helps a bit, but still doesn't test in the middle versions.

But I agree, having a flag to prefer installed, or not (or oldest; many way choice) would be helpful. Then good CI setups could test against all options. (Note to myself to add `--prefer-oldest` step to `haskell-ci`).

IMHO, the default choice doesn't really make a good argument for CI, you should test against as many "naturally occuring" dependencies as possible. Building against bundled boot libs will occur as long as `ghc` is not reinstallable.

So I'm :+1: to have more options to the solver. I actually don't care what's the default is. (If I don't like it, I hope I can change it in the global cabal-install config).

**Comment by phadej** [comment]
This is my general stance about these "let's change the default" proposals. Make options available, make them configurable, and *then* you can change the default.
 
 The recent `--enable-documentation` change *which cannot be undone* is driving me crazy.

**Comment by TeofilC** [comment]
Yes this definitely should be configurable. That really lowers the cost of changes defaults as you say.
It's especially important for this one because as people have said earlier this is load-bearing in some nix setups.

**Comment by Mikolaj** [comment]
> This is my general stance about these "let's change the default" proposals. Make options available, make them configurable, and _then_ you can (discuss the) change of the default.

I wholeheartedly agree. Given that this discussion took so long and didn't bring us consensus nor even a review of available options, I'd propose to implement the new behaviour as a flag and I hope we can merge this PR without as long a delay and afterwards let's resume discussing the default, with a very concrete and tested flag in hand that will be the candidate for the new default.

**Comment by juhp** [comment]
From the distro perspective I think the current behaviour is good, so I am not at all convinced changing the default behaviour is desirable, but having a flag is fine of course. If a dependency is already satisfied why should it have to be updated by default? Modern compiled languages already burn far too much energy with needless constant rebuilding...

**Comment by hasufell** [comment]
> From the distro perspective I think the current behaviour is good, 

That's an odd thing to say, since every distro I know has the opposite behavior.

If you install a package and there are updates for its dependencies, they will get pulled in.

----

Can I get a proper decision from Cabal team what the way forward is?

**Comment by geekosaur** [comment]
You might be interested to know that GHC HQ is considering my suggestion to ship a list of packages that can't be upgraded and possibly a list of packages that "freeze" the entire set of bootlibs, which would allow the solver to determine when it has to behave the way it does now and otherwise treat bootlibs like normal. (This would work better if ghcup could retrofit this data into older ghc versions which won't include it.) That said, @grayjay told me this may be difficult for the solver to use; it's possible that the current situation is the best the current solver can do.

**Comment by hasufell** [comment]
> This would work better if ghcup could retrofit this data into older ghc versions which won't include it.

That won't be easily possible due to https://github.com/haskell/ghcup-hs/issues/361 which is a major rewrite that has stalled

**Comment by Mikolaj** [comment]
> Can I get a proper decision from Cabal team what the way forward is?

In the absence of a PR, if you'd like the Cabal team as a whole to make a decision, please add this topic to the agenda of the fortnightly cabal devs chat at https://hackmd.io/X62yS0d6RxW3ybh8AmRqlw or, even better, please come to the meeting. Until we have such a decision or any objection, my proposal from a few comments back stands.

**Comment by hasufell** [comment]
> if you'd like the Cabal team as a whole to make a decision

Yes, please add the topic to the agenda. I don't have capacity to be involved in cabal meetings, thanks.

**Comment by Ericson2314** [comment]
And note that isn't the ideal thing for Nix either. When we hop in a dev shell we shouldn't be re-planning at all. Forcing cabal to choose installed packages is just a hack around not being able to use the original plan from which the nix shell dependencies were chosen.

**Comment by phadej** [comment]
> When we hop in a dev shell we shouldn't be re-planning at all. 

There could be greedy pick-installed-only solver for Nix (and alike) use cases. But IIRC cabal-install developers are going to remove support for different solvers (e.g. https://github.com/haskell/cabal/pull/9206), so there's no going back. (Constructing install plan, even from existing one is still "constructing install plan").

**Comment by grayjay** [comment]
> You might be interested to know that GHC HQ is considering my suggestion to ship a list of packages that can't be upgraded and possibly a list of packages that "freeze" the entire set of bootlibs, which would allow the solver to determine when it has to behave the way it does now and otherwise treat bootlibs like normal. (This would work better if ghcup could retrofit this data into older ghc versions which won't include it.) That said, @grayjay told me this may be difficult for the solver to use; it's possible that the current situation is the best the current solver can do.

@geekosaur I thought about the freezing behavior that you mentioned, and I realized that I'm not sure I understand. Would the solver need any additional logic beyond the existing requirement that an installed package depend on the exact installed versions of the dependencies that it was built against? I would expect that including `ghc` in an install plan would already freeze all of the dependencies declared in `ghc`'s `build-depends` field.

**Comment by geekosaur** [comment]
I think the issue is that, since it's a "special" package that is dependent on the exact packages shipped with ghc (more specifically, on the exact packages that `ghc` itself is built against), it can't rely on `build-depends` but on the global package db. Isn't that why bootlibs use those versions to begin with?

**Comment by grayjay** [comment]
cabal already only considers the installed version of `ghc`, because it appears in the list of non-reinstallable packages:

https://github.com/haskell/cabal/blob/63c486a1a65de599fa435c0cbf11ad85725f3c6c/cabal-install/src/Distribution/Client/Dependency.hs#L451-L476

Choosing the installed `ghc` means that all of `ghc`'s dependencies must also be installed (the same versions it was built against, which came with GHC). I should have said that cabal reads the dependencies from the `InstalledPackageInfo`, not `build-depends`, because it is an installed package. I was wondering whether cabal needs to do anything more to implement the freezing behavior that you mentioned.

**Comment by geekosaur** [comment]
It's already doing the freezing behavior; the problem is that it's freezing _too much_ and we need a way to say "only these packages need to be frozen unless the `ghc` package is involved".

**Comment by geekosaur** [comment]
To be more clear: certain packages are hardcoded into cabal as being completely non-reinstallable. Any other package in the global package db is "soft non-reinstallable", or `preferred` in cabal file lingo. They shouldn't be: they should be treated like any other package, _unless_ `ghc` is being linked. But then we need more than the version, and the global package db is not a store with a hash so we can't use that to say "this _exact_ package, not just the same version" when it comes to `ghc`. (This is not entirely true as of 9.10.1, there is a very short hash on the package id which makes the first point in the above comment false, but I don't know how it reflects into `ghc`'s dependencies.)

**Comment by TeofilC** [comment]
The issue for amending the list of non-reinstallable packages is https://github.com/haskell/cabal/issues/10087, perhaps we could move conversation related to that into that issue considering this thread is already really long

**Comment by gbaz** [comment]
an interesting tweak to cabal's design would be let the nonreinstallable package list live externally -- like have it specified by ghc, so it can more easily vary by ghc version.

**Comment by geekosaur** [comment]
I pitched that to the GHC devs some time back (and separately to @hasufell in the hope that it might be backported to older ghcs). The response at the time wasn't particularly heartening. Also @grayjay had concerns about integrating it into the solver, although the design I had proposed (this was pre-unit id hashes) could be translated easily into the equivalent of freeze files/`cabal.config` files. (Unit ID hashes complicate this because I suspect the solver needs to be taught about them somehow.)

**Comment by TeofilC** [comment]
> an interesting tweak to cabal's design would be let the nonreinstallable package list live externally -- like have it specified by ghc, so it can more easily vary by ghc version.

See: https://github.com/haskell/cabal/pull/10982, which implements something like this.

Note that this isn't what this issue is about. Changing the non-reinstallable list is https://github.com/haskell/cabal/issues/10087

**Comment by TeofilC** [comment]
I will try to tackle this by adding a configuration option to disable the current behaviour. The flag would be off by default (at least to begin with).

**Comment by hasufell** [comment]
I think this issue should remain open until the default has changed.

**Comment by Mikolaj** [comment]
Makes sense.
